[Gan Wan Cheng Isaac] iP

orzymandias opened the pull request.

Commit:
Add Gradle support

orzymandias committed:
Level-1: Greet, Echo, Exit

orzymandias committed:
Level 2. Add, List

orzymandias committed:
Level 3. Mark as Done

orzymandias committed:
Level-4: ToDo, Event, Deadline

orzymandias committed:
Handle case where user gives an input which does not contain any of the keywords.

orzymandias committed:
Automated Text UI Testing. Use the input/output redirection technique to semi-automate the testing of Intrubot.

orzymandias committed:
Teach Duke to deal with errors such as incorrect inputs entered by the user and refactored duke file

orzymandias committed:
Updated test suite to cover exception scenerios

orzymandias committed:
Add support for deleting tasks from the list.

orzymandias committed:
Updated test suite to cover deletion feature and also error handling scenerios for deletion feature.

orzymandias committed:
patch EXPECTED.TXT trailing space issue for the bot greeting, caused by Intellij stripping trailing white spaces when saving file.

orzymandias committed:
Add Save Feature to persist task history

Intrubot does not have any means to persist tasks history across different sessions.

DataSaver class introduced to handle saving and loading of task history across sessions. Modify Task class and its subclasses with secondary constructors to accept more parameters for initialization.

orzymandias committed:
Add date and times feature for Deadline and Event classes

String representation of time are not standardized and user dependent.

LocalDate used to store time in Event and Deadline classes.

Format of time is now standardized and ready for parsing and manipulation by Intrubot.

orzymandias committed:
Merge branch 'branch-Level-8'

orzymandias committed:
Merge branch 'master' into branch-Level-7

# Conflicts:
#	src/main/java/Duke.java

orzymandias committed:
Refactor the code to extract out closely related code as classes.

Duke class file is increasingly convoluted with similar logic for different function including from Ui, Storage, Command, Parser and  Tasklist

Extract logically similar code into common components into the following:
* Ui : handle reading user input and render responses
* Storage: Read and Write TaskList Items to Local Memory
* Command: Execute different logic based on command specifed
* Parser: Parse user input into their commands and arguments
*TaskLists: contains the task list

Abstract logically similar component for easier extensibility

orzymandias committed:
Divide classes into packages

Source package is increasingly cluttered and difficult to navigate

Organize the classes into suitable java packages.

Related Classes are group together to allow information hiding via package access modifier

orzymandias committed:
Add JUnit tests

Individual classes and methods are not are tested independently of other classes.

*Unit test written for Storage class under 2 scenarios loading data and reading from file if filepath is absent.
*Unit test for output stream for Ui class of  bye command

Unit testing: testing individual units ensure that event the smallest component works correctly.

orzymandias committed:
Add Find Functionality

User is unable to query items from task list.

Add FindCommand class which extract tasks containing queried term.

orzymandias committed:
Add JavaDoc comments to the code.

Public classes and methods do not have JavaDocs which hinder readability.

Add header comments to all non-private classes/methods, and non-trivial private methods.

orzymandias committed:
merge branch branch-Level-9

orzymandias committed:
Merge branch-A-JavaDoc to master

orzymandias committed:
Tweak code to follow the Coding Standard

Code does not comply with coding standard which could hinder readability and collaboration with other developers.

Coding standard has been tweaked to follow: https://se-education.org/guides/conventions/java/intermediate.html#package-and-import-statements

Allow others familiar with coding standard to understand the repository better.

orzymandias committed:
Merge branch 'master' into branch-A-CodingStandard

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/Storage.java
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/Ui.java
#	src/main/java/duke/command/ByeCommand.java
#	src/main/java/duke/command/ToDoCommand.java

orzymandias committed:
Merge branch 'master' into add-gradle-support

orzymandias committed:
Automate project builds using Gradle

Building the from source code and running testcases can be a and be tedious multi-step process.

Gradle integration saves precious development time and labour.

Gradle is being used to automate building and running of Duke. It is also set up to run the whole test suite of unit tests.

This saves the number of clicks to run duke and unit tests.

Use Gradle to automate some of the build tasks of the project.

orzymandias committed:
Use checkStyle to detect coding style violations

Checking for coding style is a manual process which is easy to miss by the developer and takes up time when it can be automated.

Automate the process of coding style checks using Gradle by adding checkstyle.xml with coding style specifications and suppressions for exceptions to ignore.

This provide specification for the checkstyle plugin to follow.

orzymandias committed:
Add a GUI to Duke

GUI is running solely in the command line with no graphical user interface.

A purely CLI interface may not be as user-friendly as a GUI interface which
more users would be accustomed to.

Use the JavaFX technology to implement the GUI. View of the application is
rendered using FXML which is manipulated using by Controller classes
MainWindow and DialogBox.

To add sub nodes individually to the main node for building the view is
harder to visualize and makes code more verbose.

ZoroarkDarkrai commented on src/main/java/duke/DialogBox.java:
```diff
@@ -0,0 +1,61 @@
+        getChildren().setAll(tmp);
+        setAlignment(Pos.TOP_LEFT);
+    }
+
```
Can add JavaDoc comment for this method?

ZoroarkDarkrai commented on src/main/java/duke/DialogBox.java:
```diff
@@ -0,0 +1,61 @@
+    public static DialogBox getUserDialog(String text, Image img) {
+        return new DialogBox(text, img);
+    }
+
```
Can add JavaDoc comment for this method?

ZoroarkDarkrai commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,61 @@
+
+    /**
+     * Instantiates Duke Object which initializes variables needed throughout program. 
+     * 
```
Can remove this extra line?

ZoroarkDarkrai commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,61 @@
+            }
+        }
+    }
+
```
Can add JavaDoc comment for this method?

ZoroarkDarkrai commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,10 @@
+ * Represents the Exception object which would be thrown throughout the process of running Duke.
+ */
+public class DukeException extends Exception {
+    public DukeException(String errorMessage) {
```
Can add JavaDoc comment for this method?

ZoroarkDarkrai commented on src/main/java/duke/Launcher.java:
```diff
@@ -0,0 +1,12 @@
+ * A launcher class to workaround classpath issues.
+ */
+public class Launcher {
+    public static void main(String[] args) {
```
Can add JavaDoc comment for this method?

ZoroarkDarkrai commented on src/main/java/duke/MainWindow.java:
```diff
@@ -0,0 +1,61 @@
+                DialogBox.getDukeDialog("Hello! I'm Duke!\nWhat can I do?", dukeImage)
+        );
+    }
+
```
Can add JavaDoc comment for this method?

ZoroarkDarkrai commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,69 @@
+     * @param reply user message input into Intrubot.
+     * @return Command to be executed by Duke class.
+     * @throws DukeException if invalid command or arguments specified.
+     */
```
Any reason why this method has the default access modifier instead of public?

ZoroarkDarkrai commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,98 @@
+    }
+    
+    /**
+     * Initializes task list object with specified Task List.
```
Can add @param in the javadoc?

ZoroarkDarkrai commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    /**
+     * Get task from task list.
```
Missing s from "Get"?

ZoroarkDarkrai commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    /**
+     * Remove task from task list.
```
Missing s from "Remove"?

ZoroarkDarkrai commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,98 @@
+
+    /**
+     * Clears all tasks from taskList.
+     * @return
```
Missing empty line?

ZoroarkDarkrai commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,125 @@
+     * @param reply to be wrapped in partition.
+     * @return String representation of reply.
+     */
+
```
Extra empty line?

ZoroarkDarkrai commented on src/main/java/duke/command/ByeCommand.java:
```diff
@@ -0,0 +1,33 @@
+     *
+     * @param taskItems represents the tasks which is added to the task list.
+     * @param ui        ui component which user interacts with or sees.
+     * @param storage   Object for saving and loading tasks list to hard disk.
```
Can add @return?

ZoroarkDarkrai commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,125 @@
+    /**
+     * Shows Task deleted message.
+     *
+     * @param task which is deleted.
```
Can add @param for task list?

ZoroarkDarkrai commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,125 @@
+    /**
+     * Shows Task added message.
+     *
+     * @param task which is added.
```
Can add @param for task list?

ZoroarkDarkrai commented on src/main/java/duke/command/ClearCommand.java:
```diff
@@ -0,0 +1,18 @@
+import duke.Storage;
+import duke.TaskList;
+import duke.Ui;
+
```
Can add Javadoc of the class?

ZoroarkDarkrai commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,28 @@
+     * 
+     * @param taskItems represents the tasks which is added to the task list.
+     * @param ui ui component which user interacts with or sees.
+     * @param storage Object for saving and loading tasks list to hard disk.
```
Can add @return?

ZoroarkDarkrai commented on src/main/java/duke/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,37 @@
+     *
+     * @param taskItems represents the tasks which is added to the task list.
+     * @param ui        ui component which user interacts with or sees.
+     * @param storage   Object for saving and loading tasks list to hard disk.
```
Can add @return?

ZoroarkDarkrai commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,56 @@
+     *
+     * @param taskItems represents the tasks which is added to the task list.
+     * @param ui        ui component which user interacts with or sees.
+     * @param storage   Object for saving and loading tasks list to hard disk.
```
Can add @return?

ZoroarkDarkrai commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,56 @@
+    private String args;
+
+    /**
+     * Initializes the DeleteCommand Object. 
```
Can add empty line?

ZoroarkDarkrai commented on src/main/java/duke/command/DoneCommand.java:
```diff
@@ -0,0 +1,57 @@
+
+
+    /**
+     * Initializes the DoneCommand Object. 
```
Can add empty line?

ZoroarkDarkrai commented on src/main/java/duke/command/DoneCommand.java:
```diff
@@ -0,0 +1,57 @@
+     *
+     * @param taskItems represents the tasks which is added to the task list.
+     * @param ui        ui component which user interacts with or sees.
+     * @param storage   Object for saving and loading tasks list to hard disk.
```
Can add @return?

ZoroarkDarkrai commented on src/main/java/duke/command/EventCommand.java:
```diff
@@ -0,0 +1,36 @@
+     *
+     * @param taskItems represents the tasks which is added to the task list.
+     * @param ui        ui component which user interacts with or sees.
+     * @param storage   Object for saving and loading tasks list to hard disk.
```
Can add @return?

ZoroarkDarkrai commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,34 @@
+
+import java.util.ArrayList;
+import java.util.List;
+
```
Can add Javadoc of the class?

ZoroarkDarkrai commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,34 @@
+
+public class FindCommand extends Command {
+    private String arg;
+    public FindCommand(String arg) {
```
Can add Javadoc for the method?

ZoroarkDarkrai commented on src/main/java/duke/command/ListCommand.java:
```diff
@@ -0,0 +1,35 @@
+     *
+     * @param taskItems represents the tasks which is added to the task list.
+     * @param ui        ui component which user interacts with or sees.
+     * @param storage   Object for saving and loading tasks list to hard disk.
```
Can add @return?

ZoroarkDarkrai commented on src/main/java/duke/command/ToDoCommand.java:
```diff
@@ -0,0 +1,51 @@
+     *
+     * @param taskItems represents the tasks which is added to the task list.
+     * @param ui        ui component which user interacts with or sees.
+     * @param storage   Object for saving and loading tasks list to hard disk.
```
Can add @return?

ZoroarkDarkrai commented on src/main/java/duke/stubs/ToDoStub.java:
```diff
@@ -0,0 +1,9 @@
+package duke.stubs;
+
+import duke.task.ToDo;
+
```
Can add javadoc for the class?

ZoroarkDarkrai commented on src/main/java/duke/stubs/ToDoStub.java:
```diff
@@ -0,0 +1,9 @@
+import duke.task.ToDo;
+
+public class ToDoStub extends ToDo {
+    public ToDoStub(String description) {
```
Can add javadoc for the method?

ZoroarkDarkrai commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,28 @@
+
+import java.time.format.DateTimeFormatter;
+import java.time.LocalDate;
+
```
Can add javadoc for the class?

ZoroarkDarkrai commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,28 @@
+        super(description);
+        this.by = by;
+    }
+
```
Can add javadoc for the method?

ZoroarkDarkrai commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,28 @@
+
+public class Deadline extends Task {
+    protected LocalDate by;
+    public Deadline(String description, LocalDate by) {
```
Can add javadoc for the class?

ZoroarkDarkrai commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,28 @@
+        super(description, isDone);
+        this.by = time;
+    }
+    
```
Can add javadoc for the class?

ZoroarkDarkrai commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,27 @@
+
+import java.time.format.DateTimeFormatter;
+import java.time.LocalDate;
+
```
Can add javadoc for the class?

ZoroarkDarkrai commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,27 @@
+
+public class Event extends Task {
+    protected LocalDate at;
+    public Event(String description, LocalDate at) {
```
Can add javadoc for the method?

ZoroarkDarkrai commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,27 @@
+        super(description);
+        this.at = at;
+    }
+    
```
Can add javadoc for the method?

ZoroarkDarkrai commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,27 @@
+    public LocalDate getAtWhen() {
+        return this.at;
+    }
+
```
Can add javadoc for the method?

ZoroarkDarkrai commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,38 @@
+package duke.task;
+
```
Can add javadoc for the class?

ZoroarkDarkrai commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,38 @@
+public class Task {
+    protected String description;
+    protected boolean isDone;
+
```
Can add javadoc for the method?

ZoroarkDarkrai commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,38 @@
+        this.description = description;
+        this.isDone = false;
+    }
+
```
Can add javadoc for the method?

ZoroarkDarkrai commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,38 @@
+        this.description = description;
+        this.isDone = isDone;
+    }
+    
```
Can add javadoc for the method?

ZoroarkDarkrai commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,38 @@
+    public String getStatusIcon() {
+        return (isDone ? "\u2713" : "\u2718"); //return tick or X symbols
+    }
+
```
Can add javadoc for the method?

ZoroarkDarkrai commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,38 @@
+    public void markDone() {
+        this.isDone = true;
+    }
+    
```
Can add javadoc for the method?

ZoroarkDarkrai commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,38 @@
+    public boolean getIsDone() {
+        return this.isDone;
+    }
+    
```
Can add javadoc for the method?

ZoroarkDarkrai commented on src/main/java/duke/task/ToDo.java:
```diff
@@ -0,0 +1,24 @@
+package duke.task;
+
+public class ToDo extends Task {
+    public ToDo(String description) {
```
Can add javadoc for the method?

ZoroarkDarkrai commented on src/main/java/duke/task/ToDo.java:
```diff
@@ -0,0 +1,24 @@
+    public ToDo(String description) {
+        super(description);
+    }
+    public ToDo(String description, boolean isDone) {
```
Can add javadoc for the method?

ZoroarkDarkrai reviewed (commented):
Nice job!

orzymandias committed:
Format code to adhere to coding standards

New code added for this week's increments were not adhering to the coding standards.

Add Javadoc's, remove a few white spaces across multiple files, reorder import statements among other adjustments.

Code is now cleaner, easier to read for people familiar with the coding standard.

vigneshbhuvan-nus commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,129 @@
+        File savedTasks = new File(filePath);
+        try {
+            if (!savedTasks.exists()) {
+                handleFileOrDirectoryDoesNotExist(savedTasks);
```
Very good piece of code, good use of abstraction. But perhaps the naming of certain methods could be a bit more short?. Hahha i'm really scraping the bell here because I can't find anything wrong

vigneshbhuvan-nus reviewed (commented):
Very good piece of code. Good job with the abstract command class and the other Command classes, really simplified understanding the code and would be very useful later on as the project evolves.

orzymandias committed:
Use assert feature  to document important assumptions throughout code

Codebase is not supported by assertions to guard against unexpected
values and state.

Implement several assertions such as but not limited to:
*Assertion is used to verify tasklist length for store operations.
*Assertion is also checks precondition for Delete command from tasklist.

Verify that code hold up to assumptions and reduce cases program state
causing unexpected bugs.

orzymandias committed:
Improve code quality

Code difficult to read due to unabstracted logic, unclear comments.

Improve the code quality by the following:
* Eliminate magic numbers and strings.
* Shorten methods like parse.
* Reduce depth of nesting in parse.
* Rename variables.
* SLAP applied to parse.
* Use default branch in conditionals.
* Rephrase comments.

Code is more readable and less error prone.

orzymandias committed:
Improve code quality even more

Code difficult to read due to unabstracted logic, unclear comments.

Improve the code quality by the following:
* Eliminate more  magic numbers and strings.

Code is more readable and less error prone.

orzymandias committed:
Merge pull request #1 from orzymandias/branch-A-Assertions

Use assert feature  to document important assumptions throughout code

orzymandias committed:
Merge branch 'master' into branch-A-CodeQuality

orzymandias committed:
Merge pull request #2 from orzymandias/branch-A-CodeQuality

Branch a code quality

orzymandias committed:
Add duplicate detection functionality

Duplicate tasks can be added to the tasklist.

Add the ability to recognize and deal with duplicate items,
by refusing addition of duplicated tasks which share the
same description and timings.

All tasks moving forward is unique and not redundant.

orzymandias committed:
Merge pull request #3 from orzymandias/branch-C-DetectDuplicates

Add duplicate detection functionality

orzymandias committed:
Set up CI

Automation by gradle is done locally.

GitHub Actions will run the gradle processes when
code is pushed or when there is a pull request.

This allows for code to be tested when pushed and
need not be fetched in order to test.

orzymandias committed:
Refactor gradle.yml to correct path

orzymandias committed:
fix syntax error in gradle.yml

orzymandias committed:
fix checkstyle errors

orzymandias committed:
Add Trivia functionality

Intrubot only stores tasks information.

User may want to store trivia data.

Store questions and answers which can
be deleted.

orzymandias committed:
:Merge branch 'branch-D-Trivia'

orzymandias committed:
Remove IO redirection test

orzymandias committed:
Merge branch 'master' of https://github.com/orzymandias/ip

orzymandias committed:
Format user guide

orzymandias committed:
Merge branch 'master' of https://github.com/orzymandias/ip

orzymandias committed:
Fix styling issues

orzymandias committed:
Fix bug for trivia

Able to save duplicate trivia and unable to search for
trivia.

*Modify trivia command to handle duplicates.
*Modify trivia class to have question and answer
in description.

*Allow trivia to be found using search
functionality.

orzymandias committed:
Add Ui.png to Documentation

orzymandias committed:
Set theme jekyll-theme-tactile

orzymandias committed:
Format user guide

orzymandias committed:
Set theme jekyll-theme-cayman

orzymandias committed:
Update user interface

Ui of GUI can be more appealing.

*Update greeting message.
*Update styling.

orzymandias committed:
Update README.md
